### PR TITLE
Change build to check in generated files

### PR DESCRIPTION
**Pull request by jaredpar**
The motivation for this change is to simplify our build and develop
steps.

Previously our compiler builds included files generated into the Obj
folder during previous phases of the build.  This meant that in order to
open Roslyn.sln successfully the code needed to be built first.  Not
ideal.

The generated code involved is rarely changed.  But in order to build
the generators themselves must be built and run as a part of every
build.  This is doable but adds unncessary overhead to development and
complicates our build files a bit.

This approach checks in the generated files which should have the
following benefits:

- Roslyn.sln can be open after a restore
- Build process is simplified and marginally faster for many operations
- Generated code now participates in GitLink

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

File: build/scripts/generate-compiler-code.ps1 [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    foreach ($fileName in (Get-ChildItem $scratchDir)) { 
+        Write-Host "Checking $fileName"
+        $realFilePath = Join-Path $generatedDir $fileName
+        $scratchFilePath = Join-Path $scratchDir $fileName
+        $scratchHash = (Get-FileHash $scratchFilePath -algorithm $algo).Hash
+        $realHash = (Get-FileHash $realFilePath -algorithm $algo).Hash
+        if ($scratchHash -ne $realHash) { 
+            Write-Host "Files are out of date"
+            Write-Host "Run $(Join-Path $PSScriptRoot generate-compiler-code.ps1) to refresh"
+            throw "Files are out of date"
```
Comment (VSadov) [inline build/scripts/generate-compiler-code.ps1]: "Out of sync with the xml definition" ?
Perhaps also include the file name in the error.

**Comment by jaredpar** [comment]
retest windows_eta_open_prtest please

File: build/scripts/generate-compiler-code.ps1 (hunk 1 above)
Comment (jaredpar) [inline build/scripts/generate-compiler-code.ps1]: The name of the file is printed just above on line 42.

**Comment by sharwell** [comment]
:question: Do the code generators meet the following properties?

* Is the output of each generator deterministic per the inputs in source control?
* For a small change in the input, is the output ever going to have a large change (e.g. will a change to member A in the file cause unrelated member B to be relocated to a different position in the file)?

**Comment by jaredpar** [comment]
@sharwell 

> Is the output of each generator deterministic per the inputs in source control?

Yes.  All aspects of our build are deterministic.  This is verified on every check in via the windows_determinism_prtest leg.  It's also central to our ability to efficiently run tests as we use content caching on our test results. 

> For a small change in the input, is the output ever going to have a large change (e.g. will a change to member A in the file cause unrelated member B to be relocated to a different position in the file)?

No.  Changes are pretty local.

File: build/Targets/GenerateCompilerInternals.targets [hunk 2]
```diff
@@ -1,177 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (agocke) [inline build/Targets/GenerateCompilerInternals.targets]: I love this change.

**Comment by agocke** [comment]
@jaredpar Admit it -- your primary motivation for making this change is to pad your GitHub diff stats.

**Comment by jaredpar** [comment]
>  Admit it -- your primary motivation for making this change is to pad your GitHub diff stats.

Now that I'm a manager I have to take every possibly opportunity.  😄 

In seriousness: i will switch that to @dotnet-bot because it needs some love.

**Pull request merged**